src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -2030,296 +1393,56 @@ ObjectMonitor* ObjectSynchronizer::inflate(Thread* self, oop object,
+                                           const char* cnt_name, size_t cnt,
+                                           LogStream* ls, elapsedTimer* timer_p) {
+  if (!SafepointMechanism::should_process(self)) {
+    return;
   }
 
-  // Sanity checks for the races:
-  guarantee(mid->owner_is_DEFLATER_MARKER(), "must be deflater marker");
-  guarantee(mid->contentions() < 0, "must be negative: contentions=%d",
-            mid->contentions());
-  guarantee(mid->_waiters == 0, "must be 0: waiters=%d", mid->_waiters);
-  guarantee(mid->_cxq == NULL, "must be no contending threads: cxq="
-            INTPTR_FORMAT, p2i(mid->_cxq));
-  guarantee(mid->_EntryList == NULL,
-            "must be no entering threads: EntryList=" INTPTR_FORMAT,
-            p2i(mid->_EntryList));
-
-  if (obj != NULL) {
-    if (log_is_enabled(Trace, monitorinflation)) {
-      ResourceMark rm;
-      log_trace(monitorinflation)("deflate_monitor: object=" INTPTR_FORMAT
-                                  ", mark=" INTPTR_FORMAT ", type='%s'",
-                                  p2i(obj), obj->mark().value(),
-                                  obj->klass()->external_name());
-    }
-
-    // Install the old mark word if nobody else has already done it.
-    mid->install_displaced_markword_in_object(obj);
+  // A safepoint/handshake has started.
+  if (ls != NULL) {
+    timer_p->stop();
+    ls->print_cr("pausing %s: %s=" SIZE_FORMAT ", in_use_list stats: ceiling="
+                 SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
+                 op_name, cnt_name, cnt, in_use_list_ceiling(),
+                 _in_use_list.count(), _in_use_list.max());
   }
-  mid->clear_common();
 
-  assert(mid->object_peek() == NULL, "must be NULL: object=" INTPTR_FORMAT,
-         p2i(mid->object_peek()));
-  assert(mid->is_free(), "must be free: allocation_state=%d",
-         (int)mid->allocation_state());
+  // Honor block request.
+  ThreadBlockInVM tbivm(self);
 
-  // Move the deflated ObjectMonitor to the working free list
-  // defined by free_head_p and free_tail_p.
-  if (*free_head_p == NULL) {
-    // First one on the list.
-    *free_head_p = mid;
-  }
-  if (*free_tail_p != NULL) {
-    // We append to the list so the caller can use mid->_next_om
-    // to fix the linkages in its context.
-    ObjectMonitor* prevtail = *free_tail_p;
-    // prevtail should have been cleaned up by the caller:
-#ifdef ASSERT
-    ObjectMonitor* l_next_om = unmarked_next(prevtail);
-    assert(l_next_om == NULL, "must be NULL: _next_om=" INTPTR_FORMAT, p2i(l_next_om));
-#endif
-    om_lock(prevtail);
-    prevtail->set_next_om(mid);  // prevtail now points to mid (and is unlocked)
+  if (ls != NULL) {
+    ls->print_cr("resuming %s: in_use_list stats: ceiling=" SIZE_FORMAT
+                 ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT, op_name,
+                 in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
+    timer_p->start();
```

Comment:
ThreadBlockInVM have a miss-feature: it safepoint polls on front-edge and back-edge.
It should only have that poll on backedge, once that is fixed, this will do the wrong thing.
Also you may safepoint on both front-edge and back-edge now, so the timer would show the wrong thing then.

So to get the timer correct you should use:
SafepointMechanism::process_if_requested(thread);

Reply:
The baseline code (ObjectSynchronizer::deflate_common_idle_monitors())
uses ThreadBlockInVM currently. I don't want to change that as part of
this work. If we want to generally change uses of ThreadBlockInVM to
something else, then we should do that with a dedicated bug.

Reply:
Currently there is no issue with ThreadBlockInVM since there is no code inside those scopes.
This adds code there which assumes the timer will be 'resumed', and logs "resume" when it actually could be going to a safepoint.

Reply:
So if I narrow the scope around the ThreadBlockInVM, then it would be fine?

```
{
  // Honor block request.
  ThreadBlockInVM tbivm(self);
}
```

Reply:
Yes that avoids it!

Reply:
Done. I also did the one in ObjectSynchronizer::request_deflate_idle_monitors().

Reply:
I like it but can you do one thing for me?  Can you s/chk/check/ in the name?

Reply:
I'd rather not. It is a function with six freaking parameters so
I went with names as short as I could...